### DISCUSSION TITLE
Migration 5.9.4 to 6.0.0-rc.2 : Problem with PostgreSQL driver or Knex ?

### DISCUSSION THREAD

**Discussion by Laryan1** [mikro-orm/mikro-orm#5018] [Q&A]
Hi ! 
It's my first time asking for help here, so I hope I'll do everything the right way.

Problem : While trying to migrate a 5.9.4 project to the last rc of v6 (I really need the "$some" query condition for Collections), I encountered many problems which I resolved following the migration guide, but I can't get a hold of this one.

I use Mikro-orm with the postgreSQL Driver, I first resolved all the issues mentionned in the migration guide. Reverting to 5.9.4, the problem disappear, and updating to 6.0.0-rc.2, it reappears.

What I found is that it might be a problem with knex where the query is expecting way more data than provided, but I didn't find how to get more logs or how to fix the issue. 

This is the bit of code that is posing problem : (other bits have the same problem, but this one is the less complex one)

```
  async updateDependentEntities(supplierProductEntity: SupplierProductEntity) {
    const contractProducts = await this.contractProductsRepository.find(
      { supplierProduct: { supplier: supplierProductEntity.supplier.id, reference: supplierProductEntity.reference } },
      {
        populate: [
          'supplierProduct.purchaseUnit',
          'supplierProduct.genericProduct',
          'contract.supplier',
          'contract.company',
        ],
      }
    );

    for await (const contractProduct of contractProducts) {
      contractProduct.updateUnitPrice();
      this.contractProductsRepository.getEntityManager().persist(contractProduct);
    }

    await this.contractProductsRepository.getEntityManager().flush();
  }
```
I found that the problem only happens on "flush".

The only logs I get from the debug of Mikro-orm concerning this is the select for the find (no problem here) and then : 

```
[Nest] 28576  - 12/12/2023 17:36:10     LOG [MikroORM] [query] begin
[Nest] 28576  - 12/12/2023 17:36:10     LOG [MikroORM] [query] rollback
```

No sql being displayed between the begin and the rollback.

Stack trace below : 
```
recette-app\node_modules\@mikro-orm\core\platforms\ExceptionConverter.js:8
      return new exceptions_1.DriverException(exception);
                 ^
DriverException: Expected 13 bindings, saw 2
    at PostgreSqlExceptionConverter.convertException (C:\...\recette-app\node_modules\@mikro-orm\core\platforms\ExceptionConverter.js:8:16)
    at PostgreSqlExceptionConverter.convertException (C:\...\recette-app\node_modules\@mikro-orm\postgresql\PostgreSqlExceptionConverter.js:42:22)
    at PostgreSqlDriver.convertException (C:\...\recette-app\node_modules\@mikro-orm\core\drivers\DatabaseDriver.js:319:54)
    at C:\...\recette-app\node_modules\@mikro-orm\core\drivers\DatabaseDriver.js:323:24
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at PostgreSqlDriver.nativeUpdate (C:\...\recette-app\node_modules\@mikro-orm\knex\AbstractSqlDriver.js:439:19)

previous Error: Expected 13 bindings, saw 2
    at replaceRawArrBindings (C:\...\recette-app\node_modules\knex\lib\formatter\rawFormatter.js:27:11)
    at Raw.toSQL (C:\...\recette-app\node_modules\knex\lib\raw.js:77:13)
    at unwrapRaw (C:\...\recette-app\node_modules\knex\lib\formatter\wrappingFormatter.js:116:19)
    at Client_PG.parameter (C:\...\recette-app\node_modules\knex\lib\client.js:409:12)
    at QueryCompiler_PG._valueClause (C:\...\recette-app\node_modules\knex\lib\query\querycompiler.js:988:21)
    at QueryCompiler_PG.whereBasic (C:\...recette-app\node_modules\knex\lib\query\querycompiler.js:1075:12)
    at QueryCompiler_PG.where (C:\...\recette-app\node_modules\knex\lib\query\querycompiler.js:584:34)
    at QueryCompiler_PG.update (C:\...\recette-app\node_modules\knex\lib\dialects\postgres\query\pg-querycompiler.js:52:25)
    at QueryCompiler_PG.toSQL (C:\...\recette-app\node_modules\knex\lib\query\querycompiler.js:75:29)
    at QueryBuilder_PostgreSQL.toSQL (C:\...\recette-app\node_modules\knex\lib\query\querybuilder.js:84:44)
``` 
Thank you in advance for your help !

**Comment by B4nan** [comment]
Try using latest dev version (note that you need to pin the version, do not use `^` version range as semantically, RC is higher than the dev version even though it's older - say thank you to the dumb npm version resolution rules, they got it wrong), there are many fixes on top of RC2. If that won't help, I need to see a [complete reproduction](https://github.com/mikro-orm/reproduction), could be a bug.

Btw what you are doing in that piece of code is something I'd call "repository abuse", you should inject entity manager instead of doing those shenanigans :] This was one of the motivations behind the removed persist/flush methods, people overuse repositories as if they were somehow special - they are not, they are just a thin layer on top of EM, which serves as an extension point. If you do not define a custom repository, there is no real point in using this abstraction altogether. The idea is to make code simpler by hiding your custom code into a custom repository, not using repositories where you should have been using EM instead.

>       this.contractProductsRepository.getEntityManager().persist(contractProduct);

And you can completely drop this line, there is no point in manually persisting managed entities, this is a no-op. The only actual place where you need to call `em.persist()` is when you create a new entity instance via constructor (when you use `em.create()` it is called automatically).

**Reply by Laryan1** [reply]
Thank you for the quick answer.

I tried using the last dev version (6.0.0-dev.260), without much success, still getting the same error/trace. I won't have much time for the coming days to try and reproduce the bug on a simpler project. I will try to in my free time on weekend.

Thank you also for the extra knowledge about repositories and the persisting of entities, I thought of EntityRepository as an easier to understand syntax so I used it for that reason only, will switch to em from now on.

**Reply by B4nan** [reply]
> I thought of EntityRepository as an easier to understand syntax so I used it for that reason only, will switch to em from now on.

It has the same API, just the `entityName` parameter is not there (as that is bound to the repository type).
